akka-http2-support/src/main/scala/akka/http/impl/engine/http2/hpack/HeaderDecompression.scala

```diff
@@ -42,13 +44,20 @@ private[http2] object HeaderDecompression extends GraphStage[FlowShape[FrameEven
       var headers = new VectorBuilder[(String, String)]
       object Receiver extends HeaderListener {
         def addHeader(name: Array[Byte], value: Array[Byte], sensitive: Boolean): Unit =
-          // TODO: optimization: use preallocated strings for well-known names, similar to
-          // what happens in HeaderParser
+          // TODO: optimization: use preallocated strings for well-known names, similar to what happens in HeaderParser
           headers += new String(name, UTF8) → new String(value, UTF8)
       }
-      decoder.decode(ByteStringInputStream(payload), Receiver)
-      decoder.endHeaderBlock() // TODO: do we have to check the result here?
-      push(eventsOut, ParsedHeadersFrame(streamId, endStream, headers.result()))
+      try {
+        decoder.decode(ByteStringInputStream(payload), Receiver)
+        decoder.endHeaderBlock() // TODO: do we have to check the result here?
+
+        push(eventsOut, ParsedHeadersFrame(streamId, endStream, headers.result()))
+      } catch {
+        case ex: IOException ⇒
+          // this is signalled by the decoder when it failed, we want to react to this by rendering a GO_AWAY frame
+          push(eventsOut, GoAwayFrame(streamId, ErrorCode.COMPRESSION_ERROR))
```

Comment:
I was toying with 2 ideas here, either we send this way and in the demux handle and "bounce it back" like in this PR. Or we make the compression stuff a full bidistage in which case it can by itself send the frame then. The second one was closer to my original branch - happy to go back to that style (didnt get there yet since rebase hell with the moving compression around PR ;-)).

So, 1 bidistage and push back directly you say? Can do.